Fix failing e2e release test

waodim opened the pull request:
<!--

Thanks for opening a pull request! Here are some tips to get everything merged smoothly:

1. Read our contributor guidelines: https://github.com/Dynatrace/dynatrace-operator/blob/main/CONTRIBUTING.md

2. If the PR is unfinished, raise it as a draft or prefix the title with "WIP:" so it's clear to everyone.

3. Be sure to allow edits from maintainers, so it's easier for us to help: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->

## Description

<!--

Please include the following:
- The motivation for the change
    - Link to the Github issue or Jira ticket, if exists.
- The summary of the change

-->

With the introduction of `v1beta2` this e2e test failed. This was because the operator was installed with using the `0.15.0` tag. This version of the operator does not know `v1beta2`, thus it failed. I fixed this by simply checking if the used version for the helm command does know `v1beta2` or not.

related ticket: https://dt-rnd.atlassian.net/browse/K8S-10110

## How can this be tested?

run `make test/e2e/release`

<!--

Please include some guiding steps on how to test this change during a review.

- What environment is necessary for the change to be noticeable ?
- What needs to be enabled/applied for the change to be noticeable ?

-->

andriisoldatenko commented on test/scenarios/release/release_test.go:
```diff
@@ -5,22 +5,27 @@ package release
 var testEnv env.Environment
+var thresholdVersion, _ = dtversion.ToSemver("1.1.0")
+
+const usedVersion = "0.15.0"
```
hmm... any reason why do we use old version for running e2e tests?

chrismuellner replied:
We should bump this version as the test should just test the upgrade from previous to latest version

andriisoldatenko replied:
☝️  that was my point as well

waodim replied:
Fair enough. I will bump it to 1.0.0 then I guess?

andriisoldatenko commented on test/features/cloudnative/codemodules/codemodules.go:
```diff
@@ -98,7 +98,7 @@ func InstallFromImage(t *testing.T) features.Feature {
 
 	// Register dynakube install
-	dynakube.Install(builder, helpers.LevelAssess, &secretConfigs[0], cloudNativeDynakube)
+	dynakube.Install(builder, helpers.LevelAssess, &secretConfigs[0], cloudNativeDynakube, false)
```
I don't understand this change, if we never use `true`, why do we need it?

waodim replied:
it is used in the release e2e test, where I check if we use a old version which does not know v1beta2

waodim replied:
I am not sure how to do that differently. In the case we use a version that does not know v1beta2 we convert our dynakube to v1beta1 and use v1beta1 then instead of v1beta2

wepudt reviewed (dismissed):
tests pass, but the solution will have to be adapted based on the version we use and whether this version uses CRD v1beta1 or v1beta2, I guess.

0sewa0 commented on test/helpers/components/dynakube/dynakube.go:
```diff
@@ -27,9 +29,16 @@ const (
-func Install(builder *features.FeatureBuilder, level features.Level, secretConfig *tenant.Secret, testDynakube dynakubev1beta2.DynaKube) {
-	Create(builder, level, secretConfig, testDynakube)
-	VerifyStartup(builder, level, testDynakube)
+func Install(builder *features.FeatureBuilder, level features.Level, secretConfig *tenant.Secret, testDynakube dynakubev1beta2.DynaKube, usesOldVersion bool) {
+	if usesOldVersion {
+		dkv1beta1 := dynakubev1beta1.DynaKube{}
+		dkv1beta1.ConvertFrom(&testDynakube)
+		CreateV1Beta1(builder, level, secretConfig, dkv1beta1)
+		VerifyStartupV1Beta1(builder, level, dkv1beta1)
+	} else {
+		Create(builder, level, secretConfig, testDynakube)
+		VerifyStartup(builder, level, testDynakube)
+	}
 }
```
you could just introduce an `InstallAsOld` or something, so you only change a single package and not every

```
func InstallAsOld(builder *features.FeatureBuilder, level features.Level, secretConfig *tenant.Secret, testDynakube dynakubev1beta2.DynaKube) {
	dkv1beta1 := dynakubev1beta1.DynaKube{}
	dkv1beta1.ConvertFrom(&testDynakube)
	CreateV1Beta1(builder, level, secretConfig, dkv1beta1)
	VerifyStartupV1Beta1(builder, level, dkv1beta1)
}
```

The name `InstallAsOld` is not my greatest, so pls try to change it 😅

waodim replied:
I had this idea also yesterday in the evening, on it :+1:

waodim replied:
done in https://github.com/Dynatrace/dynatrace-operator/pull/3205/commits/0ce7b62c82226527c1db1832f2fb808d8d81983a

Please give me feedback for the naming 😄

waodim commented:
@0sewa0 @andriisoldatenko @wepudt 

Guys I did some changes:

- i removed the releaseTag out of the helm install command, i did this because then the helm command takes the latest stable released version and installs the operator
- I added logic to get our latest tag from the operator and check if this version knows v1beta2
- I used 1.2.0 as threshold version, as this will be the first version that knows v1beta2.

Please give me feedback what you think about those changes.

@chrismuellner gave me the hint that helm always gets the latest version when no specific version is set.

0sewa0 commented on test/scenarios/release/release_test.go:
```diff
@@ -30,8 +34,19 @@ func TestMain(m *testing.M) {
 }
 
 func TestRelease(t *testing.T) {
+	cmd := exec.Command("git", "describe", "--tags", "--abbrev=0")
+	latestTag, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Fatalf("Failed to run 'git describe': %v", err)
+	}
+
+	usesOldVersion := false
+	usedSemVer, _ := dtversion.ToSemver(string(latestTag))
+	if semver.Compare(thresholdVersion, usedSemVer) >= 1 {
+		usesOldVersion = true
+	}
```
will this work in the CI?

if we are still running the things on Concourse in the background then we do not have `git` when we run this.
- Concourse just gets the content of the branch/tag, but doesn't do a full `git clone`, so we can't use git
  - If we could use `git` then we would have done this from the beginning

andriisoldatenko replied:
As we run it using the new Github Action Runner maybe you try to read it from envy or so.

@0sewa0 wdyt if we just bump and hardcode version again?

0sewa0 replied:
as this supposed to be a bugfix, I would actually prefer it, we should revisit the "get the version dynamically" problem in a separate ticket

andriisoldatenko replied:
@waodim  ☝️

waodim replied:
Got it. I will go back to the previous approach.

waodim replied:
done https://github.com/Dynatrace/dynatrace-operator/pull/3205/commits/d771f4da34f9527c3661b0cfb31d886a11db4bbf

0sewa0 commented on test/scenarios/release/release_test.go:
```diff
@@ -30,8 +35,13 @@ func TestMain(m *testing.M) {
 }
 
 func TestRelease(t *testing.T) {
+	usesOldVersion := false
+	usedSemVer, _ := dtversion.ToSemver(releaseTag)
+	if semver.Compare(thresholdVersion, usedSemVer) >= 1 {
+		usesOldVersion = true
+	}
```
we are comparing basically 2 `const` here, where the outcome is always the same, why are we doing this? 😅 

is there a case where the "upgrade" test scenario will become a "downgrade" test scenario? (I hope not 😅 )

can we determine the `usesOldVersion` is any nice way?
- meaning: will there be some check that will magically tell us if version A supports the new CR version that version B uses?
  - I don't think so, or at least not now

waodim replied:
My thought process was: currently we are at `1.1.0`, which does not know `v1beta2`. When `1.2.0` is released we have a version that knows `v1beta2`. From then on (after we bump the `releaseTag` in the test manually, because it is hardcoded) we always use the normal install method. After that the here introduced changes become useless and may be removed. But at the moment I do not see a other way to fix this test. What do you think about this? Do you have some tips what I could change here?

0sewa0 replied:
There are multiple perspectives:
1. **Does it hurt to always check if the conversion works properly?** (ie. just always consider `usesOldVersion` as `true`)
  - IMO, it should be fine to check if the conversion is working properly
    - kinda mixes the responsibilities of the tests, which is not nice, but not horrible 
  - this results in a "dumber" logic when it comes to the e2e test code (we don't have extra checks)
2. **Automate it to the fullest:**
  - So we never have to manually update any values, it "just works" and tests the upgrade from `latest released version` to the `current branch`.
  - Your check would kinda make sense in that situation
  - For this to work the following would need to be true:
     1. We (dynamically) know what is the `latest released version`
     2. We (dynamically) know what is the `latest CR version` that the `latest released version` supports
     3. We (dynamically) know what is the `latest CR version` that the `current branch` supports
     4. We (dynamically) know what is the "oldVersion" to use to convert to 🙃 
     5. ...
3. **Partially automate it**
 - So we only have to intervene, when a new version is introduced 
 - This is what you did
   - More accurately, "would have done", if the `git` magic worked 😅 


I do not think option 2. is anyway feasible.

I like option 1, for its simplicity, but can be annoying to keep up to date

Option 3 would make sense and would be the best "middle ground" IF the `git` magic can be used.
- As of right now we do not now
- Introducing code that "might be used in the future" is something we should avoid, as it will only cause confusion.
  - If I asked anyone, 2 weeks from now, "Why do we have that code?", the likely answer would be "dunno, does not make much sense, but don't want to touch it as it works, right?"

waodim replied:
I guess the most practical way currently is to go for option 1, then. I will go for it. Thank you for elaborating.

0sewa0 replied:
np, as mentioned before, we should look into **option 3.** separately  😉

waodim replied:
done in https://github.com/Dynatrace/dynatrace-operator/pull/3205/commits/ba567e6920800e3d0d4ab60466bb2464fe1e8390

I removed the whole logic with the check if we use an old version because if we assume it is always true we do not need that flag.
